rclrs/src/node/primitive_options.rs

```diff
@@ -0,0 +1,253 @@
+    fn transient_local(self) -> PrimitiveOptions<'a> {
+        self.durability(QoSDurabilityPolicy::TransientLocal)
+    }
+
+    /// Override the default [`QoSProfile::lifespan`] for the primitive.
+    fn lifespan(self, lifespan: QoSDuration) -> PrimitiveOptions<'a> {
+        let mut options = self.into_primitive_options();
+        options.lifespan = Some(lifespan);
+        options
+    }
+
+    /// Set a custom duration for the [lifespan][QoSProfile::lifespan] of the primitive.
+    fn lifespan_duration(self, duration: Duration) -> PrimitiveOptions<'a> {
+        self.lifespan(QoSDuration::Custom(duration))
+    }
+
+    /// Make the [lifespan][QoSProfile::lifespan] of the primitive infinite.
+    fn infinite_lifespan(self) -> PrimitiveOptions<'a> {
+        self.lifespan(QoSDuration::Infinite)
+    }
+
+    /// Override the default [`QoSProfile::deadline`] for the primitive.
+    fn deadline(self, deadline: QoSDuration) -> PrimitiveOptions<'a> {
+        let mut options = self.into_primitive_options();
+        options.deadline = Some(deadline);
+        options
+    }
+
+    /// Set the [`QoSProfile::deadline`] to a custom finite value.
+    fn deadline_duration(self, duration: Duration) -> PrimitiveOptions<'a> {
+        self.deadline(QoSDuration::Custom(duration))
+    }
+
+    /// Do not use a deadline for liveliness for this primitive.
+    fn no_deadline(self) -> PrimitiveOptions<'a> {
+        self.deadline(QoSDuration::Infinite)
+    }
+
+    /// Override the default [`QoSProfile::liveliness_lease`] for the primitive.
+    fn liveliness_lease(self, lease: QoSDuration) -> PrimitiveOptions<'a> {
+        let mut options = self.into_primitive_options();
+        options.liveliness_lease = Some(lease);
+        options
+    }
+
+    /// Set a custom duration for the [liveliness lease][QoSProfile::liveliness_lease].
+    fn liveliness_lease_duration(self, duration: Duration) -> PrimitiveOptions<'a> {
+        self.liveliness_lease(QoSDuration::Custom(duration))
+    }
+
+    /// [Avoid the ROS namespace conventions][1] for the primitive.
+    ///
+    /// [1]: QoSProfile::avoid_ros_namespace_conventions
+    fn avoid_ros_namespace_conventions(self) -> PrimitiveOptions<'a> {
+        let mut options = self.into_primitive_options();
+        options.avoid_ros_namespace_conventions = Some(true);
+        options
+    }
+}
+
+impl<'a> IntoPrimitiveOptions<'a> for PrimitiveOptions<'a> {
+    fn into_primitive_options(self) -> PrimitiveOptions<'a> {
+        self
+    }
+}
+
+impl<'a> IntoPrimitiveOptions<'a> for &'a str {
+    fn into_primitive_options(self) -> PrimitiveOptions<'a> {
+        PrimitiveOptions::new(self)
+    }
+}
+
+impl<'a, T: Borrow<str>> IntoPrimitiveOptions<'a> for &'a T {
+    fn into_primitive_options(self) -> PrimitiveOptions<'a> {
+        self.borrow().into_primitive_options()
+    }
+}
+
+impl<'a> PrimitiveOptions<'a> {
```

Comment:
I would strongly discourage this for the following specific reason:

The idea of a "default" set of values for QoS is a bad one in my opinion. If it were entirely up to me, `QOS_PROFILE_DEFAULT` would not exist in rcl, and we would not provide `QoSProfile::default()`. The reason is, it should not be assumed that a single set of QoS settings would make the best default settings across both topics and services.

Currently `QOS_PROFILE_DEFAULT` and `QOS_PROFILE_SERVICES_DEFAULT` happen to have the same set of values, and `QOS_PROFILE_DEFAULT` is used as the default for topics, but by parading around `QOS_PROFILE_DEFAULT` as "the sensible default" across both topics and services, rcl has locked itself into hurting users if there's ever a shift in opinion about what the best default settings are for topics, and those settings aren't good for services, because there are probably many users out there who are using `QOS_PROFILE_DEFAULT` for services right now. Personally I think rcl should deprecate `QOS_PROFILE_DEFAULT` and replace it with `QOS_PROFILE_TOPICS_DEFAULT`.

That being said, maybe I'm being overly sensitive about this.

Reply:
> Simplifying a bit, I personally would expect A.apply(B) to update or modify A, but in this API its the opposite.

Would it help if the API were expressed as `PrimitiveOptions::apply_to(&mut QoSProfile)`?

I could also accept `QoSProfile::apply(&mut PrimitiveOptions)`, but personally I'd rather the `qos` module not know anything about the `primitive_options` module since I'd rather have a simple clean dependency of `primitive_options` depends on `qos` rather than a circular dependency where they both depend on each other. Rust has no issue supporting circular dependencies between modules that are in the same crate, but it feels like bad hygiene to me.

Reply:
Yeah I think we can just rename apply to `apply_to`, that'll work

Reply:
Updated here: [a872cd6](https://github.com/ros2-rust/ros2_rust/pull/429/commits/a872cd6c8eee2919909cc2e7771cbe2bd97881cd)